src/PushThread.cpp

```diff
@@ -38,1149 +38,893 @@ static const char *RcsId = "$Header: /home/cvsadm/cvsroot/fermi/servers/hdb++/hd
-#endif
-		{
-			string ret = signals[i].dberror;
-			sig_lock->unlock();
-			return ret;
-		}
-	}
-
-	sig_lock->unlock();
-	return STATUS_DB_ERROR;
-}
-//=============================================================================
-/**
- *	Increment the error counter of db saving
- */
-//=============================================================================
-void  PushThreadShared::set_nok_db(string &signame, string error)
-{
-	sig_lock->lock();
-	unsigned int i;
-	for (i=0 ; i<signals.size() ; i++)
-	{
-		if (signals[i].name==signame)
-		{
-			signals[i].nokdb_counter++;
-			signals[i].nokdb_counter_freq++;
-			signals[i].dbstate = Tango::ALARM;
-			signals[i].dberror = STATUS_DB_ERROR;
-			if(error.length() > 0)
-				signals[i].dberror += ": " + error;
-			gettimeofday(&signals[i].last_nokdb, NULL);
-			sig_lock->unlock();
-			return;
-		}
-	}
-	for (i=0 ; i<signals.size() ; i++)
-	{
-#ifndef _MULTI_TANGO_HOST
-		if (hdb_dev->compare_without_domain(signals[i].name,signame))
-#else
-		if (!hdb_dev->compare_tango_names(signals[i].name,signame))
+        try
+        {
+            (hdb_dev->_device)->push_change_event("AttributePendingNumber",&hdb_dev->AttributePendingNumber);
+            (hdb_dev->_device)->push_archive_event("AttributePendingNumber",&hdb_dev->AttributePendingNumber);
+        }
+        catch(Tango::DevFailed &e)
+        {
+
+        }
 #endif
-		{
-			signals[i].nokdb_counter++;
-			signals[i].nokdb_counter_freq++;
-			signals[i].dbstate = Tango::ALARM;
-			signals[i].dberror = STATUS_DB_ERROR;
-			if(error.length() > 0)
-				signals[i].dberror += ": " + error;
-			gettimeofday(&signals[i].last_nokdb, NULL);
-			sig_lock->unlock();
-			return;
-		}
-	}
-	if(i == signals.size())
-	{
-		HdbStat sig;
-		sig.name = signame;
-		sig.nokdb_counter = 1;
-		sig.nokdb_counter_freq = 1;
-		sig.okdb_counter = 0;
-		sig.store_time_avg = 0.0;
-		sig.store_time_min = -1;
-		sig.store_time_max = -1;
-		sig.process_time_avg = 0.0;
-		sig.process_time_min = -1;
-		sig.process_time_max = -1;
-		sig.dbstate = Tango::ALARM;
-		sig.dberror = STATUS_DB_ERROR;
-		if(error.length() > 0)
-			sig.dberror += ": " + error;
-		gettimeofday(&sig.last_nokdb, NULL);
-		signals.push_back(sig);
-	}
-	sig_lock->unlock();
-}
-//=============================================================================
-/**
- *	Get the error counter of db saving
- */
-//=============================================================================
-uint32_t  PushThreadShared::get_nok_db(string &signame)
-{
-	sig_lock->lock();
-	for (unsigned int i=0 ; i<signals.size() ; i++)
-	{
-		if (signals[i].name==signame)
-		{
-			uint32_t ret = signals[i].nokdb_counter;
-			sig_lock->unlock();
-			return ret;
-		}
-	}
-	for (unsigned int i=0 ; i<signals.size() ; i++)
-	{
+        std::vector<std::shared_ptr<HdbCmdData>> cmds;
+        for(size_t i = 0; i < events_size; ++i)
+        {
+            cmds.push_back(events.front());
+            events.pop_front();
+        }
+
```

Comment:
@dlacoste-esrf  I think here hdb_dev->AttributePendingNumber should be set to 0 or to events.size() (we're inside the lock, so values different than 0 shouldn't be possible), otherwise it will always be >= 1.
Do you agree?
Better ideas?

Reply:
This is still missing in my opinion to merge this PR.

Reply:
you're completely right, I forgot to look it up last time. It is done. What about the commented code for pushing events ? is it needed ?

Reply:
It would have been nice to push events just when the value of AttributePendingNumber changes, but at some point it was deadlocking, so it was moved in the PollerThread.
To re-enable it, all combinations of involved mutexes and Tango Monitor should be checked.

Reply:
Indeed… Actually the pollerThread only role is to push events… The same logic could be applied to all of them and we could get rid of it…
Maybe we should open an issue to do that on another PR, what do you think ?